scene-lib/src/scenelib/annotations/el/TypePathEntry.java

```diff
@@ -0,0 +1,169 @@
+package scenelib.annotations.el;
+
+import org.objectweb.asm.TypePath;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A TypePathEntry is a way to get from one node in a {@link TypePath} to another.
+ * One can treat these as edges in a graph.
+ *
+ * ASM has a data structure {@link TypePath}.  {@code List<TypePathEntry>}
+ * corresponds to a {@code TypePath}.  That is, each TypePathEntry corresponds
+ * to a step in an ASM TypePath.
+
+ * {@code List<TypePathEntry>} also corresponds to the javac class
+ * {@code com.sun.tools.javac.code.TypeAnnotationPosition}.
+ */
+public class TypePathEntry {
+  /**
+   * The kind of TypePathEntry; that is, how to get from the previous node in a TypePath to this one.
+   * One of TypePath.ARRAY_ELEMENT, TypePath.INNER_TYPE, TypePath.WILDCARD_BOUND, TypePath.TYPE_ARGUMENT.
+   *
+   * This corresponds to javac class {@code com.sun.tools.javac.code.TypeAnnotationPosition.TypePathEntry}.
+   */
+  public final int step;
+  /**
+   * If this represents a type argument (that is, step == TYPE_ARGUMENT, then the index for the type argument.
+   * Otherwise, 0.
+   */
+  public final int argument;
+
+  /**
+   * Construct a new TypePathEntry.
+   *
+   * @param step the type of the TypePathEntry
+   * @param argument index of the type argument or 0
+   */
+  public TypePathEntry(int step, int argument) {
+    this.step = step;
+    this.argument = argument;
+
+    assert step == TypePath.ARRAY_ELEMENT || step == TypePath.INNER_TYPE
+        || step == TypePath.WILDCARD_BOUND || step == TypePath.TYPE_ARGUMENT;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    TypePathEntry that = (TypePathEntry) o;
+    return step == that.step && argument == that.argument;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(step, argument);
+  }
+
+  /**
+   * Converts a type path represented by a list of integers to a {@link TypePath}.
+   * @param integerList the integer list in the form [step1, argument1, step2, argument2, ...] where step1 and argument1
+   *                   are the step and argument of the first entry (or edge) of a type path.
+   * @return the {@link TypePath} corresponding with the <code>integerList</code>.
+   */
+  public static TypePath getTypePathFromBinary(List<Integer> integerList) {
+    if (integerList == null) {
+      return null;
+    }
+    StringBuilder stringBuilder = new StringBuilder();
+    Iterator<Integer> iterator = integerList.iterator();
+    while (iterator.hasNext()) {
+      int step = iterator.next();
+      if (!iterator.hasNext()) {
+        throw new IllegalArgumentException("Could not decode type path: " + integerList);
+      }
+      int argument = iterator.next();
+      switch (step) {
+        case TypePath.ARRAY_ELEMENT:
```

Comment:
Consider abstracting out a `stepToString` method that could be used here and one other place.